### ISSUE TITLE
Hypothesis doesn't recognize unittest.SkipTest exception

### ISSUE THREAD

**Issue by coriolinus** [HypothesisWorks/hypothesis-python#514]
# Narrative example of failure

I haven't had time to construct a minimal example, so I'll give a full example of failure. I've recently updated a test in my Django project which conditionally skips a test in cases which no longer make sense. 

The test in question is defined in an abstract base class of a couple other test classes which specialize it further. One branch of one test no longer makes sense in the context of one of the specialization classes. I can't solve this with `assume()` or `@unittest.skipIf()` because neither has enough contextual data. Instead, I `raise unittest.SkipTest()` at runtime if the test should be skipped.

The issue is that Hypothesis doesn't know about `unittest.SkipTest`, so produces failure output, as seen below:

```shell
(.venv) coriolinus$ ./manage.py test
Creating test database for alias 'default'...
Installed 39 object(s) from 1 fixture(s)
Installed 21 object(s) from 1 fixture(s)
Installed 9579 object(s) from 1 fixture(s)
Installed 5 object(s) from 1 fixture(s)
Installed 6 object(s) from 1 fixture(s)
.............................s.............................Falsifying example: test_admin_sees_everyone(self=<core.tests.test_user_profile_visibility.ProfileVisibilityTests testMethod=test_admin_sees_everyone>, data=data(...))
Draw 1: <User: alpha>
s..........................................
----------------------------------------------------------------------
Ran 102 tests in 58.985s

OK (skipped=2)
Destroying test database for alias 'default'...
(.venv) coriolinus$ pip list | grep hypothesis
hypothesis (3.6.1)
```

# Expected behavior

Hypothesis does not produce failure output if a test raises `unittest.SkipTest`.

# Observed behavior

Hypothesis produces failure output if a test raises `unittest.SkipTest`.

**Comment by alexwlchan** [comment]
Thanks for the report! I had a quick go at trying to write an example – is this sort of like what you were doing?

```python
import unittest

from hypothesis import given
from hypothesis.strategies import integers


class DemoTest(unittest.TestCase):

    @given(integers())
    def test_to_be_skipped(self, x):
        if x == 0:
            raise unittest.SkipTest()
        else:
            assert x == 0


unittest.main()
```

```console
$ python issue514.py
Falsifying example: test_to_be_skipped(self=<__main__.DemoTest testMethod=test_to_be_skipped>, x=0)
s
----------------------------------------------------------------------
Ran 1 test in 0.021s

OK (skipped=1)
```

Sounds like we should just recognise `unittest.SkipTest` as a special exception – I get the same result whether I raise the exception directly or call `self.skipTest(reason)`. Not sure how hard that is off the top of my head.

**Comment by coriolinus** [comment]
Yes, that's the essence of the issue. Agreed with the proposed course of
action. I feel like I should have submitted a PR to go with this bug
report, but I don't know anything about Hypothesis' internals.

**Comment by alexwlchan** [comment]
So at a very quick glance, I think we need to make a change somewhere around [line&nbsp;464 of core.py](https://github.com/HypothesisWorks/hypothesis-python/blob/master/src/hypothesis/core.py#L464) so that if we catch a `unittest.SkipTest` exception, we abort the entire test run.

I’m not sure what mechanisms we have for marking a test as skipped (as opposed to just silently dropping the error) – I’ll dig in a little further when I have more time.

**Comment by DRMacIver** [comment]
> I’m not sure what mechanisms we have for marking a test as skipped (as opposed to just silently dropping the error)

I think just immediately reraising the error should have the desired effect shouldn't it?

**Comment by alexwlchan** [comment]
> I think just immediately reraising the error should have the desired effect shouldn't it?

Oh yeah. PR incoming. :)

**Comment by alexwlchan** [comment]
@coriolinus This took longer than I was expecting to finish, but the fix is now released as Hypothesis 3.13.

Thanks for your report!

**Comment by coriolinus** [comment]
I'm very glad to hear it! Thanks for fixing this.
